modules/frontend/combiner/common.go

```diff
@@ -129,15 +129,25 @@ func (c *genericCombiner[T]) HTTPFinal() (*http.Response, error) {
 		return nil, err
 	}
 
-	bodyString, err := new(jsonpb.Marshaler).MarshalToString(final)
-	if err != nil {
-		return nil, fmt.Errorf("error marshalling response body: %w", err)
+	var bodyString string
+	if c.marshallingFormat == api.HeaderAcceptProtobuf {
+		buff, err := proto.Marshal(final)
+		if err != nil {
+			return nil, fmt.Errorf("error marshalling response body: %w", err)
+		}
+		bodyString = string(buff)
+	} else {
+		c.marshallingFormat = api.HeaderAcceptJSON
+		bodyString, err = new(jsonpb.Marshaler).MarshalToString(final)
+		if err != nil {
+			return nil, fmt.Errorf("error marshalling response body: %w", err)
+		}
 	}
 
 	return &http.Response{
 		StatusCode: c.httpStatusCode,
```

Comment:
it's silly we set this to 200 in all the combiners and then return it here. we should hardcode this to return 200 and only set `c.httpStatusCode` if there's an error

the combiner was brought together from ~3 different sources when i refactored the frontend. it created some warts like this b/c i wanted to change as little code as possible

this (and the above comment) suggest we really need a `newCombiner(foo &genericCombiner)` method that properly initializes these fields

Reply:
I'm really on the fence on this. Especially because you don't need to initiate the status code to make it work. In the first iteration, it won't stop and it will be replaced by the response's one. But it's "magic". I've added the method to make it more explicit